### PR TITLE
Fix relocating of moved tracks in external collections

### PR DESCRIPTION
**Pull request by uklotzde**
The update of moved tracks in external collections didn't work as expected. Only the aoide integration in #2282 was affected affected. The *old id* of the old track and the *new location* of the new track need to be merged into a new `TrackRef` that identifies the relocated track.

This bug was mainly caused by using a `QPair` instead of a custom class. The class `RelocatedTrack` now encapsulates all the information with appropriate accessor functions and lots of assertions to check consistency. I also added checks to verify that Track::getTrackAdded() is valid for tracks in the library. This was the first indicator for the wrong behavior.

**Note 1**: Relinking library directories in Mixxx still requires to rescan the library manually. This should be fixed independently. Another bug is that the track table view is not updated with the new location after rescanning. This might also affect BaseTrackCache, I didn't check.

**Note 2**: ~~During testing duplicate tracks appeared in aoide. This is a different issue that I need to check.~~ Not reproducible, might be fixed. Aoide v0.5.0 will prevent this actively.

### CODE REVIEW

**Comment by uklotzde** [comment]
Rebased on master since no comments received and only a single commit.

**Review by daschuer — CHANGES_REQUESTED** [review]
I had expected that after using Relink in Library preferences all tracks will come back without scanning the library. This is not the case. Is this a bug? 

* create folder 1 with one track
* start Mixxx add the folder in the library preferences 
* Play the track 
* quit Mixxx 
* rename the folder to 2
* start Mixxx go to the library preferences  
* Relink folder 1 to 2 
* try to play the track 
* ... Missing 
* Rescan Library track is back

**Comment by uklotzde** [comment]
> I had expected that after using Relink in Library preferences all tracks will come back without scanning the library. This is not the case. Is this a bug?
> 
>     * create folder 1 with one track
> 
>     * start Mixxx add the folder in the library preferences
> 
>     * Play the track
> 
>     * quit Mixxx
> 
>     * rename the folder to 2
> 
>     * start Mixxx go to the library preferences
> 
>     * Relink folder 1 to 2
> 
>     * try to play the track
> 
>     * ... Missing
> 
>     * Rescan Library track is back

This is a known bug that I discovered while fixing another bug. I was sure that I have already reported it, but I haven't found it yet. Out of scope, will not be fixed in this PR.

**Comment by uklotzde** [comment]
> > I had expected that after using Relink in Library preferences all tracks will come back without scanning the library. This is not the case. Is this a bug?
> > ```
> > * create folder 1 with one track
> > 
> > * start Mixxx add the folder in the library preferences
> > 
> > * Play the track
> > 
> > * quit Mixxx
> > 
> > * rename the folder to 2
> > 
> > * start Mixxx go to the library preferences
> > 
> > * Relink folder 1 to 2
> > 
> > * try to play the track
> > 
> > * ... Missing
> > 
> > * Rescan Library track is back
> > ```
> 
> This is a known bug that I discovered while fixing another bug. I was sure that I have already reported it, but I haven't found it yet. Out of scope, will not be fixed in this PR.

Ups, I mentioned it right here in this PR, see above *Note 1* :see_no_evil: We should create a bug report.

**Comment by uklotzde** [comment]
Rescan after relinking: https://bugs.launchpad.net/mixxx/+bug/1862396

File: src/library/relocatedtrack.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        // new track to the internal collection before.
+        DEBUG_ASSERT(m_addedTrackRef.hasLocation());
+        DEBUG_ASSERT(mergedTrackRef().isValid());
+        DEBUG_ASSERT(mergedTrackRef().hasId());
+        DEBUG_ASSERT(mergedTrackRef().hasLocation());
+    }
+
+    // The new TrackRef of the relocated track after merging
+    // the missing with the newly added track.
+    TrackRef mergedTrackRef() const {
```
[the remark this answers is not in the document]
Comment (uklotzde) [reply src/library/relocatedtrack.h]: updatedTrackRef() might better reflect the naming at the call side?

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
LGTM, Thank you.

**Pull request merged**